cpp/benchmarks/tpch/utils.hpp

```diff
@@ -28,17 +30,104 @@
+                                           "l_commitdate",
+                                           "l_receiptdate",
+                                           "l_returnflag",
+                                           "l_linestatus",
+                                           "l_shipinstruct",
+                                           "l_shipmode",
+                                           "l_comment",
+                                           "l_extendedprice"};
+const std::vector<std::string> PART     = {"p_partkey",
+                                           "p_name",
+                                           "p_mfgr",
+                                           "p_brand",
+                                           "p_type",
+                                           "p_size",
+                                           "p_container",
+                                           "p_retailprice",
+                                           "p_comment"};
+const std::vector<std::string> PARTSUPP = {
+  "ps_partkey", "ps_suppkey", "ps_availqty", "ps_supplycost", "ps_comment"};
+const std::vector<std::string> SUPPLIER = {
+  "s_suppkey", "s_name", "s_address", "s_nationkey", "s_phone", "s_acctbal", "s_comment"};
+const std::vector<std::string> CUSTOMER = {"c_custkey",
+                                           "c_name",
+                                           "c_address",
+                                           "c_nationkey",
+                                           "c_phone",
+                                           "c_acctbal",
+                                           "c_mktsegment",
+                                           "c_comment"};
+const std::vector<std::string> NATION   = {"n_nationkey", "n_name", "n_regionkey", "n_comment"};
+const std::vector<std::string> REGION   = {"r_regionkey", "r_name", "r_comment"};
+
+}  // namespace
+
+/**
+ * @brief Log the peak memory usage of the GPU
+ */
+class memory_stats_logger {
+ public:
+  memory_stats_logger()
+    : existing_mr(rmm::mr::get_current_device_resource()),
+      statistics_mr(rmm::mr::make_statistics_adaptor(existing_mr))
+  {
+    rmm::mr::set_current_device_resource(&statistics_mr);
+  }
+
+  ~memory_stats_logger() { rmm::mr::set_current_device_resource(existing_mr); }
+
+  [[nodiscard]] void print_peak_memory_usage() const noexcept
+  {
+    std::cout << "Peak memory usage: "
+              << static_cast<double>(statistics_mr.get_bytes_counter().peak / (1024 * 1024))
+              << " MB" << std::endl;
+  }
+
+ private:
+  rmm::mr::device_memory_resource* existing_mr;
+  rmm::mr::statistics_resource_adaptor<rmm::mr::device_memory_resource> statistics_mr;
+};
+
 // RMM memory resource creation utilities
 inline auto make_cuda() { return std::make_shared<rmm::mr::cuda_memory_resource>(); }
+inline auto make_async()
+{
+  return std::make_shared<rmm::mr::cuda_async_memory_resource>(
+    rmm::percent_of_free_device_memory(50));
```

Comment:
How did we choose 50%? It seems like we could provide no argument and use the RMM default, which is also 50%. Reference: [RMM docs for cuda_async_memory_resource](https://docs.rapids.ai/api/rmm/nightly/librmm_docs/memory_resources/#_CPPv4N3rmm2mr26cuda_async_memory_resource26cuda_async_memory_resourceENSt8optionalINSt6size_tEEENSt8optionalINSt6size_tEEENSt8optionalI22allocation_handle_typeEE)